In [ ]:
import polars as pl
from scipy.stats import chi2_contingency
import numpy as np

In [2]:
df = pl.read_csv("Capstone2025_nsi_lvl9_with_landcover_and_color.csv")
df

shape: (2_417_766, 8)
┌─────────────────┬──────┬───────────┬──────────┬────────────┬───────────────────┬────────┬────────┐
│ h3              ┆ fips ┆ st_damcat ┆ bldgtype ┆ lc_type    ┆ loc               ┆ clr    ┆ clr_cc │
│ ---             ┆ ---  ┆ ---       ┆ ---      ┆ ---        ┆ ---               ┆ ---    ┆ ---    │
│ str             ┆ i64  ┆ str       ┆ str      ┆ str        ┆ str               ┆ str    ┆ i64    │
╞═════════════════╪══════╪═══════════╪══════════╪════════════╪═══════════════════╪════════╪════════╡
│ 8929a960d6fffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.828147 ┆ orange ┆ 8      │
│                 ┆      ┆           ┆          ┆            ┆ 398273 37.072…    ┆        ┆        │
│ 8929a978c07ffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.634281 ┆ orange ┆ 8      │
│                 ┆      ┆           ┆          ┆            ┆ 56983 36.9838…    ┆        ┆        │
│ 892836da147ffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.464731 ┆ orange ┆ 11     │
│                 ┆      ┆           ┆          ┆            ┆ 172229 37.323…    ┆        ┆        │
│ 892836c644fffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.787252 ┆ orange ┆ 12     │
│                 ┆      ┆           ┆          ┆            ┆ 301526 37.429…    ┆        ┆        │
│ 892836d736fffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.716239 ┆ orange ┆ 1      │
│                 ┆      ┆           ┆          ┆            ┆ 893409 37.339…    ┆        ┆        │
│ …               ┆ …    ┆ …         ┆ …        ┆ …          ┆ …                 ┆ …      ┆ …      │
│ 892836da15bffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.457547 ┆ olive  ┆ 6      │
│                 ┆      ┆           ┆          ┆            ┆ 536701 37.320…    ┆        ┆        │
│ 892836c226bffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.591661 ┆ olive  ┆ 1      │
│                 ┆      ┆           ┆          ┆            ┆ 571591 37.401…    ┆        ┆        │
│ 892836d121bffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.492292 ┆ olive  ┆ 1      │
│                 ┆      ┆           ┆          ┆            ┆ 055735 37.305…    ┆        ┆        │
│ 892836dae97ffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.479228 ┆ olive  ┆ 2      │
│                 ┆      ┆           ┆          ┆            ┆ 840586 37.332…    ┆        ┆        │
│ 892836d1923ffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.603759 ┆ olive  ┆ 4      │
│                 ┆      ┆           ┆          ┆            ┆ 523727 37.352…    ┆        ┆        │
└─────────────────┴──────┴───────────┴──────────┴────────────┴───────────────────┴────────┴────────┘

In [10]:
df.filter(pl.col("loc") == df[0, "loc"])

shape: (16, 8)
┌─────────────────┬──────┬───────────┬──────────┬────────────┬─────────────────┬──────────┬────────┐
│ h3              ┆ fips ┆ st_damcat ┆ bldgtype ┆ lc_type    ┆ loc             ┆ clr      ┆ clr_cc │
│ ---             ┆ ---  ┆ ---       ┆ ---      ┆ ---        ┆ ---             ┆ ---      ┆ ---    │
│ str             ┆ i64  ┆ str       ┆ str      ┆ str        ┆ str             ┆ str      ┆ i64    │
╞═════════════════╪══════╪═══════════╪══════════╪════════════╪═════════════════╪══════════╪════════╡
│ 8929a960d6fffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.8281 ┆ orange   ┆ 8      │
│                 ┆      ┆           ┆          ┆            ┆ 47398273        ┆          ┆        │
│                 ┆      ┆           ┆          ┆            ┆ 37.072…         ┆          ┆        │
│ 8929a960d6fffff ┆ 6047 ┆ RES       ┆ W        ┆ urban+crop ┆ POINT(-120.8281 ┆ coffee   ┆ 20     │
│                 ┆      ┆           ┆          ┆            ┆ 47398273        ┆          ┆        │
│                 ┆      ┆           ┆          ┆            ┆ 37.072…         ┆          ┆        │
│ 8929a960d6fffff ┆ 6047 ┆ RES       ┆ W        ┆ urban+crop ┆ POINT(-120.8281 ┆ crimson  ┆ 5      │
│                 ┆      ┆           ┆          ┆            ┆ 47398273        ┆          ┆        │
│                 ┆      ┆           ┆          ┆            ┆ 37.072…         ┆          ┆        │
│ 8929a960d6fffff ┆ 6047 ┆ RES       ┆ W        ┆ urban+crop ┆ POINT(-120.8281 ┆ gold     ┆ 1      │
│                 ┆      ┆           ┆          ┆            ┆ 47398273        ┆          ┆        │
│                 ┆      ┆           ┆          ┆            ┆ 37.072…         ┆          ┆        │
│ 8929a960d6fffff ┆ 6047 ┆ RES       ┆ W        ┆ urban+crop ┆ POINT(-120.8281 ┆ gray     ┆ 6      │
│                 ┆      ┆           ┆          ┆            ┆ 47398273        ┆          ┆        │
│                 ┆      ┆           ┆          ┆            ┆ 37.072…         ┆          ┆        │
│ …               ┆ …    ┆ …         ┆ …        ┆ …          ┆ …               ┆ …        ┆ …      │
│ 8929a960d6fffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.8281 ┆ gold     ┆ 1      │
│                 ┆      ┆           ┆          ┆            ┆ 47398273        ┆          ┆        │
│                 ┆      ┆           ┆          ┆            ┆ 37.072…         ┆          ┆        │
│ 8929a960d6fffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.8281 ┆ gray     ┆ 4      │
│                 ┆      ┆           ┆          ┆            ┆ 47398273        ┆          ┆        │
│                 ┆      ┆           ┆          ┆            ┆ 37.072…         ┆          ┆        │
│ 8929a960d6fffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.8281 ┆ lavender ┆ 7      │
│                 ┆      ┆           ┆          ┆            ┆ 47398273        ┆          ┆        │
│                 ┆      ┆           ┆          ┆            ┆ 37.072…         ┆          ┆        │
│ 8929a960d6fffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.8281 ┆ navy     ┆ 1      │
│                 ┆      ┆           ┆          ┆            ┆ 47398273        ┆          ┆        │
│                 ┆      ┆           ┆          ┆            ┆ 37.072…         ┆          ┆        │
│ 8929a960d6fffff ┆ 6047 ┆ RES       ┆ M        ┆ urban+crop ┆ POINT(-120.8281 ┆ olive    ┆ 4      │
│                 ┆      ┆           ┆          ┆            ┆ 47398273        ┆          ┆        │
│                 ┆      ┆           ┆          ┆            ┆ 37.072…         ┆          ┆        │
└─────────────────┴──────┴───────────┴──────────┴────────────┴─────────────────┴──────────┴────────┘

In [39]:
df["clr"].value_counts(sort=True)

shape: (38, 2)
┌───────────┬────────┐
│ clr       ┆ count  │
│ ---       ┆ ---    │
│ str       ┆ u32    │
╞═══════════╪════════╡
│ cocoa     ┆ 273275 │
│ alabaster ┆ 189116 │
│ amber     ┆ 153476 │
│ red       ┆ 151898 │
│ orange    ┆ 148188 │
│ …         ┆ …      │
│ scarlet   ┆ 2664   │
│ plum      ┆ 2510   │
│ tan       ┆ 2493   │
│ emerald   ┆ 955    │
│ maroon    ┆ 874    │
└───────────┴────────┘

In [40]:
df["st_damcat"].value_counts(sort=True)

shape: (4, 2)
┌───────────┬─────────┐
│ st_damcat ┆ count   │
│ ---       ┆ ---     │
│ str       ┆ u32     │
╞═══════════╪═════════╡
│ RES       ┆ 2175329 │
│ COM       ┆ 196812  │
│ IND       ┆ 30881   │
│ PUB       ┆ 14744   │
└───────────┴─────────┘

In [41]:
df["bldgtype"].value_counts(sort=True)

shape: (5, 2)
┌──────────┬─────────┐
│ bldgtype ┆ count   │
│ ---      ┆ ---     │
│ str      ┆ u32     │
╞══════════╪═════════╡
│ W        ┆ 1089774 │
│ M        ┆ 1043693 │
│ C        ┆ 127382  │
│ H        ┆ 96991   │
│ S        ┆ 59926   │
└──────────┴─────────┘

In [42]:
df["clr_cc"].describe()

shape: (9, 2)
┌────────────┬────────────┐
│ statistic  ┆ value      │
│ ---        ┆ ---        │
│ str        ┆ f64        │
╞════════════╪════════════╡
│ count      ┆ 2.417766e6 │
│ null_count ┆ 0.0        │
│ mean       ┆ 4.023648   │
│ std        ┆ 4.631857   │
│ min        ┆ 0.0        │
│ 25%        ┆ 1.0        │
│ 50%        ┆ 2.0        │
│ 75%        ┆ 5.0        │
│ max        ┆ 377.0      │
└────────────┴────────────┘

In [43]:
lc_clr = df.group_by(["lc_type", "clr"]).len()
lc_clr = (
    lc_clr
    .with_columns((pl.col("len") / pl.col("len").sum().over("lc_type")).alias("pct"))
    .sort(["lc_type", "len"], descending=[False, True])
)
lc_clr

shape: (419, 4)
┌─────────────┬───────────┬──────┬──────────┐
│ lc_type     ┆ clr       ┆ len  ┆ pct      │
│ ---         ┆ ---       ┆ ---  ┆ ---      │
│ str         ┆ str       ┆ u32  ┆ f64      │
╞═════════════╪═══════════╪══════╪══════════╡
│ barren      ┆ cocoa     ┆ 4737 ┆ 0.324674 │
│ barren      ┆ orange    ┆ 2043 ┆ 0.140027 │
│ barren      ┆ navy      ┆ 1767 ┆ 0.12111  │
│ barren      ┆ alabaster ┆ 1583 ┆ 0.108499 │
│ barren      ┆ red       ┆ 1183 ┆ 0.081083 │
│ …           ┆ …         ┆ …    ┆ …        │
│ urban+shrub ┆ scarlet   ┆ 41   ┆ 0.000388 │
│ urban+shrub ┆ auburn    ┆ 36   ┆ 0.000341 │
│ urban+shrub ┆ maroon    ┆ 36   ┆ 0.000341 │
│ urban+shrub ┆ bar       ┆ 35   ┆ 0.000331 │
│ urban+shrub ┆ emerald   ┆ 4    ┆ 0.000038 │
└─────────────┴───────────┴──────┴──────────┘

In [44]:
fips_clr = df.group_by(["fips", "clr"]).len()
fips_clr = (
    fips_clr
    .with_columns((pl.col("len") / pl.col("len").sum().over("fips")).alias("pct"))
    .sort(["fips", "len"], descending=[False, True])
)
fips_clr

shape: (716, 4)
┌──────┬────────┬───────┬──────────┐
│ fips ┆ clr    ┆ len   ┆ pct      │
│ ---  ┆ ---    ┆ ---   ┆ ---      │
│ i64  ┆ str    ┆ u32   ┆ f64      │
╞══════╪════════╪═══════╪══════════╡
│ 6001 ┆ cocoa  ┆ 14601 ┆ 0.179488 │
│ 6001 ┆ orange ┆ 11178 ┆ 0.13741  │
│ 6001 ┆ red    ┆ 9589  ┆ 0.117876 │
│ 6001 ┆ green  ┆ 8886  ┆ 0.109234 │
│ 6001 ┆ purple ┆ 8823  ┆ 0.10846  │
│ …    ┆ …      ┆ …     ┆ …        │
│ 6115 ┆ blue   ┆ 340   ┆ 0.045625 │
│ 6115 ┆ yellow ┆ 317   ┆ 0.042539 │
│ 6115 ┆ beige  ┆ 171   ┆ 0.022947 │
│ 6115 ┆ plum   ┆ 57    ┆ 0.007649 │
│ 6115 ┆ bar    ┆ 51    ┆ 0.006844 │
└──────┴────────┴───────┴──────────┘

# investigate colors

## 1. clr_cc outliers (IQR)

In [45]:
q1 = df["clr_cc"].quantile(0.25)
q3 = df["clr_cc"].quantile(0.75)
iqr = q3 - q1
upper_bound = q3 + 1.5 * iqr

f"Q1: {q1}, Q3: {q3}, IQR: {iqr}, Upper bound: {upper_bound}"

'Q1: 1.0, Q3: 5.0, IQR: 4.0, Upper bound: 11.0'

In [46]:
clr_cc_outliers = df.filter(pl.col("clr_cc") > upper_bound)
f"Outliers: {len(clr_cc_outliers):,} rows ({100 * len(clr_cc_outliers) / len(df):.2f}%)"

'Outliers: 140,828 rows (5.82%)'

In [47]:
clr_cc_outliers["clr_cc"].describe()

shape: (9, 2)
┌────────────┬───────────┐
│ statistic  ┆ value     │
│ ---        ┆ ---       │
│ str        ┆ f64       │
╞════════════╪═══════════╡
│ count      ┆ 140828.0  │
│ null_count ┆ 0.0       │
│ mean       ┆ 17.794998 │
│ std        ┆ 8.015995  │
│ min        ┆ 12.0      │
│ 25%        ┆ 13.0      │
│ 50%        ┆ 15.0      │
│ 75%        ┆ 20.0      │
│ max        ┆ 377.0     │
└────────────┴───────────┘

### Findings: clr_cc outliers

- **Upper bound**: 11 (Q3 + 1.5×IQR)
- **Outliers**: 140,828 rows
- Outlier distribution: median=15, mean=17.8, max=377

Some H3 cells have unusually high color counts. These could represent dense urban areas or data outliers?

## 2. Chi-square residuals for (lc_type, clr)

In [48]:
contingency = df.group_by(["lc_type", "clr"]).len().pivot(on="clr", index="lc_type", values="len").fill_null(0)
lc_types = contingency["lc_type"].to_list()
clr_cols = [c for c in contingency.columns if c != "lc_type"]
observed = contingency.select(clr_cols).to_numpy()
chi2, p, dof, expected = chi2_contingency(observed)
residuals = (observed - expected) / np.sqrt(expected)
residuals_df = pl.DataFrame({"lc_type": lc_types})
for i, col in enumerate(clr_cols):
    residuals_df = residuals_df.with_columns(pl.Series(name=col, values=residuals[:, i]))

f"Chi-square: {chi2:,.0f}, p-value: {p:.2e}, dof: {dof}"

'Chi-square: 701,632, p-value: 0.00e+00, dof: 444'

In [49]:
residuals_df

shape: (13, 39)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ lc_type   ┆ purple    ┆ alabaster ┆ emerald   ┆ … ┆ crimson   ┆ verde     ┆ amber     ┆ scarlet  │
│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---       ┆ ---      │
│ str       ┆ f64       ┆ f64       ┆ f64       ┆   ┆ f64       ┆ f64       ┆ f64       ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ urban+bar ┆ 76.938022 ┆ 27.639068 ┆ -5.306904 ┆ … ┆ -12.38481 ┆ -21.82950 ┆ 53.71211  ┆ -9.17992 │
│ ren       ┆           ┆           ┆           ┆   ┆           ┆ 7         ┆           ┆ 8        │
│ urban+shr ┆ 26.278021 ┆ -1.547656 ┆ -5.839379 ┆ … ┆ -8.607987 ┆ -17.95431 ┆ 12.959708 ┆ -6.98645 │
│ ub        ┆           ┆           ┆           ┆   ┆           ┆ 9         ┆           ┆ 7        │
│ shrub     ┆ -48.01828 ┆ 35.642739 ┆ -4.354083 ┆ … ┆ -9.22902  ┆ 3.117551  ┆ -36.83856 ┆ -9.38899 │
│           ┆ 2         ┆           ┆           ┆   ┆           ┆           ┆ 8         ┆ 1        │
│ urban+oth ┆ 0.627564  ┆ -0.490032 ┆ -0.25374  ┆ … ┆ -0.555066 ┆ -0.926841 ┆ 0.82477   ┆ -0.42379 │
│ er        ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ 3        │
│ crop      ┆ -19.26588 ┆ 66.392611 ┆ -0.363367 ┆ … ┆ -6.597495 ┆ -8.111642 ┆ -31.79906 ┆ -4.04457 │
│           ┆ 8         ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …        │
│ urban+for ┆ -54.62009 ┆ 2.451139  ┆ 13.650244 ┆ … ┆ 79.819791 ┆ 56.071025 ┆ -9.839303 ┆ -5.63402 │
│ est       ┆ 8         ┆           ┆           ┆   ┆           ┆           ┆           ┆ 7        │
│ urban     ┆ 54.426525 ┆ -65.43783 ┆ -19.31150 ┆ … ┆ -45.21864 ┆ -76.95469 ┆ 15.87877  ┆ 15.54172 │
│           ┆           ┆           ┆ 3         ┆   ┆ 2         ┆ 5         ┆           ┆ 7        │
│ barren    ┆ -12.31068 ┆ 13.077406 ┆ -1.150933 ┆ … ┆ -5.251442 ┆ -8.768783 ┆ -3.882346 ┆ -4.00947 │
│           ┆ 4         ┆           ┆           ┆   ┆           ┆           ┆           ┆ 6        │
│ urban+cro ┆ -5.843691 ┆ 2.408868  ┆ 20.543904 ┆ … ┆ 75.022291 ┆ -5.235834 ┆ -1.294027 ┆ 23.11189 │
│ p         ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │
│ other     ┆ -0.299364 ┆ -0.395523 ┆ -0.028107 ┆ … ┆ -0.061485 ┆ -0.102666 ┆ -0.35631  ┆ -0.04694 │
│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ 3        │
└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘

### Combinations rarer than expected under independence

In [59]:
residuals_long = residuals_df.unpivot(index="lc_type", variable_name="clr", value_name="std_residual")
residuals_long.sort("std_residual").head(10)

shape: (10, 3)
┌─────────┬──────────┬──────────────┐
│ lc_type ┆ clr      ┆ std_residual │
│ ---     ┆ ---      ┆ ---          │
│ str     ┆ str      ┆ f64          │
╞═════════╪══════════╪══════════════╡
│ urban   ┆ brown    ┆ -179.073234  │
│ urban   ┆ coffee   ┆ -108.900751  │
│ urban   ┆ lavender ┆ -101.903694  │
│ forest  ┆ red      ┆ -92.983638   │
│ forest  ┆ amber    ┆ -85.357421   │
│ forest  ┆ lemon    ┆ -85.084624   │
│ urban   ┆ sage     ┆ -82.93993    │
│ forest  ┆ purple   ┆ -78.906308   │
│ forest  ┆ foo      ┆ -77.493607   │
│ urban   ┆ verde    ┆ -76.954695   │
└─────────┴──────────┴──────────────┘

### Combinations more common than expected under independence

In [60]:
residuals_long.sort("std_residual", descending=True).head(10)

shape: (10, 3)
┌──────────────┬────────────┬──────────────┐
│ lc_type      ┆ clr        ┆ std_residual │
│ ---          ┆ ---        ┆ ---          │
│ str          ┆ str        ┆ f64          │
╞══════════════╪════════════╪══════════════╡
│ forest       ┆ coffee     ┆ 217.869055   │
│ forest       ┆ brown      ┆ 205.57155    │
│ forest       ┆ verde      ┆ 160.04574    │
│ urban+crop   ┆ yellow     ┆ 147.951568   │
│ forest       ┆ olive      ┆ 130.97014    │
│ forest       ┆ sage       ┆ 110.32688    │
│ urban        ┆ lemon      ┆ 107.400987   │
│ urban+forest ┆ lavender   ┆ 101.192939   │
│ urban+crop   ┆ aquamarine ┆ 98.918179    │
│ crop         ┆ gray       ┆ 95.004132    │
└──────────────┴────────────┴──────────────┘

Chi-square test: p ≈ 0 — color and land cover are strongly dependent (not independent).

Forest areas associated with coffee, brown, verde, olive, sage and not really red, amber, lemon. Urban areas avoid earth tones are more likely to be lemon. So synthetic color assignment is related to land cover type.